src/cryptonote_core/blockchain.h

```diff
@@ -1162,22 +1163,73 @@ class Blockchain {
     void flush_invalid_blocks();
 
     /**
-     * @brief Get the list of nodes that are removable from the network by creating a diff comparing
-     * the nodes that are present in the smart contract but _not_ in the Oxen service node list.
+     * @brief returns true if the given pubkey is removable (according to oxend) from the L2
+     * contract.
      *
-     * @return all the service nodes bls keys that should be removed from the smart contract
+     * To be removable, the node must either be:
      *
-     * TODO: This API is awkward right now because we generate a full-diff of nodes to remove but we
-     * only use this in an RPC call that selects 1 BLS public key at a time to aggregate a signature
-     * for removal. However we were interested in providing a batch API where the network aggregates
-     * a signature over a batch of nodes to liquidate. This would reduce churn but this also makes
-     * this function in its current form more appropriate for the problem we're trying to solve
-     * sensible to call for that purpose.
+     * 1) In the oxend recently removed nodes list.  These are nodes that the oxend network has
+     *    removed from service, either by failing requirements and getting deregistered, or by
+     *    successfully reaching the end of a requested unlock period.
      *
-     * This TODO is a reminder that we should utilise the full output of this function instead of
-     * using it in a single-object-at-a-time mindset.
+     * 2) Nodes that are in the contract but neither in the current registered service node list nor
+     *    with an incoming but not-yet-confirmed new service node registration event working its way
+     *    through the Oxen chain.  This is not a normal case: it indicates that some network
+     *    failure, bug, or partially invalid registration (such as an invalid Ed25519 signature)
+     *    resulting in the entry missing from oxend's service node list.  Note that this case relies
+     *    on infrequently updated cached contract service node pubkey lists; it will not reliably
+     *    include a node until at least an hour has passed (or, more precisely,
+     *    `netconf.L2_NODE_LIST_PURGE_BLOCKS` L2 blocks) since the event that triggered it (but in
+     *    general this is okay because the contract also requires a 2 hour post-registration delay
+     *    before a liquidation signature will be accepted).
```

Comment:
This was resolved out-of-band, in summary:

> liquidation:
> never allowed within 2h of registration
> requires either:
> BLS liquidation signature from SN network
> 30d (or whatever the timeout) after an exit request was initiated (and hasn't been acted on).
>
> exit requires all of:
> at least 2h since registration
> SN network exit signature
> exit request must have been initiated at least (exit period) ago